pallet-chainlink-feed/src/lib.rs

```diff
@@ -0,0 +1,1367 @@
+			let mut round = first_to_prune;
+			while round < keep_round {
+				Rounds::<T>::remove(feed_id, round);
+				Details::<T>::remove(feed_id, round);
+				round += RoundId::one();
+			}
+			feed.first_valid_round = Some(keep_round.max(first_valid_round));
+
+			Feeds::<T>::insert(feed_id, feed);
+
+			Ok(())
+		}
+
+		// --- feed: round requests ---
+
+		/// Set requester permissions for `requester`.
+		/// Limited to the feed owner.
+		#[weight = T::WeightInfo::set_requester()]
+		pub fn set_requester(
+			origin,
+			feed_id: T::FeedId,
+			requester: T::AccountId,
+			delay: RoundId,
+		) -> DispatchResultWithPostInfo {
+			let owner = ensure_signed(origin)?;
+			let feed = Self::feed_config(feed_id).ok_or(Error::<T>::FeedNotFound)?;
+			ensure!(feed.owner == owner, Error::<T>::NotFeedOwner);
+
+			// Keep the `last_started_round` if the requester already existed.
+			let mut requester_meta = Self::requester(feed_id, &requester).unwrap_or_default();
+			requester_meta.delay = delay;
+			Requesters::<T>::insert(feed_id, &requester, requester_meta);
+
+			Self::deposit_event(RawEvent::RequesterPermissionsSet(feed_id, requester, true, delay));
+
+			Ok(().into())
+		}
+
+		/// Remove requester permissions for `requester`.
+		/// Limited to the feed owner.
+		#[weight = T::WeightInfo::remove_requester()]
+		pub fn remove_requester(
+			origin,
+			feed_id: T::FeedId,
+			requester: T::AccountId,
+		) -> DispatchResultWithPostInfo {
+			let owner = ensure_signed(origin)?;
+			let feed = Self::feed_config(feed_id).ok_or(Error::<T>::FeedNotFound)?;
+			ensure!(feed.owner == owner, Error::<T>::NotFeedOwner);
+
+			let requester_meta = Requesters::<T>::take(feed_id, &requester).ok_or(Error::<T>::RequesterNotFound)?;
+
+			Self::deposit_event(RawEvent::RequesterPermissionsSet(feed_id, requester, false, requester_meta.delay));
+
+			Ok(().into())
+		}
+
+		/// Request the start of a new oracle round.
+		/// Limited to accounts with "requester" permission.
+		#[weight = T::WeightInfo::request_new_round()]
+		pub fn request_new_round(
+			origin,
+			feed_id: T::FeedId,
+		) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin)?;
+			let mut requester = Self::requester(feed_id, &sender).ok_or(Error::<T>::NotAuthorizedRequester)?;
+
+			with_transaction_result(|| -> DispatchResultWithPostInfo {
+				let mut feed = Feed::<T>::load_from(feed_id).ok_or(Error::<T>::FeedNotFound)?;
+
+				let new_round = feed.reporting_round_id().checked_add(One::one()).ok_or(Error::<T>::Overflow)?;
+				let last_started = requester.last_started_round.unwrap_or(Zero::zero());
+				let next_allowed_round = last_started.checked_add(requester.delay).ok_or(Error::<T>::Overflow)?;
+				ensure!(requester.last_started_round.is_none() || new_round > next_allowed_round, Error::<T>::CannotRequestRoundYet);
+
+				requester.last_started_round = Some(new_round);
+				Requesters::<T>::insert(feed_id, &sender, requester);
+
+				feed.request_new_round(sender)?;
+
```

Comment:
Is it possible to remove the `with_transaction_result` if you swap the previous two lines? So the `insert` does not happen if `request_new_round` fails.

Reply:
It might be possible to massage the code that way, but I think the control flow is too complex to make sure that it stays that way through refactorings, so I will leave it.